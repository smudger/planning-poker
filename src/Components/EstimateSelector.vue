<script setup>
import { RadioGroup, RadioGroupLabel, RadioGroupOption } from "@headlessui/vue";

defineProps({
  options: { type: Array, required: true },
});

defineEmits(["change"]);
</script>

<template>
  <RadioGroup
    class="mt-2"
    name="estimate"
    @update:model-value="$emit('change', $event)"
  >
    <RadioGroupLabel class="sr-only">Estimate</RadioGroupLabel>
    <div class="grid grid-cols-3 gap-3 sm:grid-cols-6">
      <RadioGroupOption
        v-for="option in options"
        :key="option"
        v-slot="{ active, checked }"
        :value="option"
        as="template"
      >
        <div
          :class="[
            'cursor-pointer focus:outline-none',
            active ? 'ring-2 ring-indigo-600 ring-offset-2' : '',
            checked
              ? 'bg-indigo-600 text-white hover:bg-indigo-500'
              : 'ring-1 ring-inset ring-gray-300 bg-white text-gray-900 hover:bg-gray-50',
            'flex items-center justify-center rounded-md py-3 px-3 text-sm font-semibold uppercase sm:flex-1',
          ]"
        >
          <RadioGroupLabel as="span">{{ option }}</RadioGroupLabel>
        </div>
      </RadioGroupOption>
    </div>
  </RadioGroup>
</template>
