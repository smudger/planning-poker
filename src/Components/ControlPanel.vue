<script setup>
import EstimateSelector from "@/Components/EstimateSelector.vue";

defineProps({
  gameState: { type: String, default: undefined },
  estimateOptions: { type: Array, required: true },
});

defineEmits(["estimate-changed", "reveal-cards", "reset-game"]);

const copyGameLinkToClipboard = () => {
  navigator.clipboard.writeText(window.location.href);
};
</script>
<template>
  <div
    class="fixed bottom-0 inset-x-0 flex items-center sm:items-end sm:justify-center flex-col sm:flex-row bg-gray-50 p-4 space-y-2 sm:space-y-0"
  >
    <div v-if="gameState === 'estimating'" class="max-w-full sm:mr-4">
      <EstimateSelector
        :options="estimateOptions"
        @change="$emit('estimate-changed', $event)"
      />
    </div>
    <div
      class="flex sm:flex-col w-full sm:w-auto space-x-4 sm:space-x-0 sm:gap-y-4 justify-center sm:justify-end"
    >
      <button
        class="w-1/2 sm:w-32 rounded-md bg-indigo-50 px-3.5 py-2.5 text-sm font-semibold text-indigo-600 shadow-sm hover:bg-indigo-100"
        data-test="copy-game-link"
        type="button"
        @click="copyGameLinkToClipboard"
      >
        Copy Link
      </button>
      <div class="w-1/2 sm:w-32 sm:mb-2">
        <button
          v-if="gameState === 'estimating'"
          class="w-full rounded-md bg-indigo-600 px-3.5 py-2.5 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
          data-test="reveal-cards"
          type="button"
          @click="$emit('reveal-cards')"
        >
          Reveal
        </button>
        <button
          v-if="gameState === 'showing'"
          class="w-full rounded-md bg-indigo-600 px-3.5 py-2.5 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
          data-test="reset-game"
          type="button"
          @click="$emit('reset-game')"
        >
          Reset
        </button>
      </div>
    </div>
  </div>
</template>
