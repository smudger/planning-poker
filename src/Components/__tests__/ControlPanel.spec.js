import { describe, expect, it } from "vitest";
import { shallowMount } from "@vue/test-utils";
import ControlPanel from "@/Components/ControlPanel.vue";
import EstimateSelector from "@/Components/EstimateSelector.vue";

describe("ControlPanel", () => {
  it('shows the estimate selector when the game state is "estimating"', () => {
    const wrapper = shallowMount(ControlPanel, {
      attachTo: document.body,
      props: {
        gameState: "estimating",
        estimateOptions: ["1", "2", "3"],
      },
    });

    expect(wrapper.findComponent(EstimateSelector).isVisible()).toBe(true);
  });

  it('does not show the estimate selector when the game state is "showing"', () => {
    const wrapper = shallowMount(ControlPanel, {
      attachTo: document.body,
      props: {
        gameState: "showing",
        estimateOptions: ["1", "2", "3"],
      },
    });

    expect(wrapper.findComponent(EstimateSelector).isVisible()).toBe(false);
  });

  it("emits the estimate changed event from when the estimate changes", async () => {
    const wrapper = shallowMount(ControlPanel, {
      props: {
        gameState: "estimating",
        estimateOptions: ["1", "2", "3"],
      },
    });

    await wrapper.findComponent(EstimateSelector).vm.$emit("change", "1");

    expect(wrapper.emitted()).toHaveProperty("estimate-changed");
    expect(wrapper.emitted("estimate-changed")).toEqual([["1"]]);
  });

  it("passes the options to the estimate selector", () => {
    const estimateOptions = ["1", "2", "3"];

    const wrapper = shallowMount(ControlPanel, {
      props: {
        gameState: "estimating",
        estimateOptions,
      },
    });

    expect(wrapper.findComponent(EstimateSelector).props("options")).toEqual(
      estimateOptions,
    );
  });

  it("displays a reveal cards button when the game state is estimating", () => {
    const wrapper = shallowMount(ControlPanel, {
      props: {
        gameState: "estimating",
        estimateOptions: ["1", "2", "3"],
      },
    });

    expect(wrapper.find('button[data-test="reveal-cards"]').exists()).toBe(
      true,
    );
  });

  it("does not display a reveal cards button when the game state is showing", () => {
    const wrapper = shallowMount(ControlPanel, {
      props: {
        gameState: "showing",
        estimateOptions: ["1", "2", "3"],
      },
    });

    expect(wrapper.find('button[data-test="reveal-cards"]').exists()).toBe(
      false,
    );
  });

  it("emits a reveal cards event when the reveal cards button is clicked", async () => {
    const wrapper = shallowMount(ControlPanel, {
      props: {
        gameState: "estimating",
        estimateOptions: ["1", "2", "3"],
      },
    });

    await wrapper.find('button[data-test="reveal-cards"]').trigger("click");

    expect(wrapper.emitted()).toHaveProperty("reveal-cards");
  });

  it("displays a reset game button when the game state is showing", () => {
    const wrapper = shallowMount(ControlPanel, {
      props: {
        gameState: "showing",
        estimateOptions: ["1", "2", "3"],
      },
    });

    expect(wrapper.find('button[data-test="reset-game"]').exists()).toBe(true);
  });

  it("does not display a reset game button when the game state is estimating", () => {
    const wrapper = shallowMount(ControlPanel, {
      props: {
        gameState: "estimating",
        estimateOptions: ["1", "2", "3"],
      },
    });

    expect(wrapper.find('button[data-test="reset-game"]').exists()).toBe(false);
  });

  it("emits a reset game event when the reset game button is clicked", async () => {
    const wrapper = shallowMount(ControlPanel, {
      props: {
        gameState: "showing",
        estimateOptions: ["1", "2", "3"],
      },
    });

    await wrapper.find('button[data-test="reset-game"]').trigger("click");

    expect(wrapper.emitted()).toHaveProperty("reset-game");
  });
});
