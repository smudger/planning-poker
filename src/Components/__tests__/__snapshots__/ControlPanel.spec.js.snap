// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ControlPanel > renders correctly 1`] = `
<div
  class="fixed bottom-0 inset-x-0 flex items-center sm:items-end sm:justify-center flex-col sm:flex-row bg-gray-50 p-4 space-y-2 sm:space-y-0"
>
  <div
    class="max-w-full sm:mr-4"
  >
    <estimate-selector-stub
      options="1,2,3"
    />
  </div>
  <div
    class="flex sm:flex-col w-full sm:w-auto space-x-4 sm:space-x-0 sm:gap-y-4 justify-center sm:justify-end"
  >
    <button
      class="w-1/2 sm:w-32 rounded-md bg-indigo-50 px-3.5 py-2.5 text-sm font-semibold text-indigo-600 shadow-sm hover:bg-indigo-100"
      data-test="copy-game-link"
      type="button"
    >
       Copy Link 
    </button>
    <div
      class="w-1/2 sm:w-32 sm:mb-2"
    >
      <button
        class="w-full rounded-md bg-indigo-600 px-3.5 py-2.5 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
        data-test="reveal-cards"
        type="button"
      >
         Reveal 
      </button>
      <!--v-if-->
    </div>
  </div>
</div>
`;
